File: algorithms/medium/cheapest_flights_within_k_stops.ex
defmodule Solution do
  @spec find_cheapest_price(
          n :: integer,
          flights :: [[integer]],
          src :: integer,
          dst :: integer,
          k :: integer
        ) :: integer
  def find_cheapest_price(n, flights, src, dst, k) do
    graph =
      flights
      |> Enum.reduce(%{}, fn [from, to, price], acc ->
        x = {to, price}
        Map.update(acc, from, [x], &[x | &1])
      end)

    :queue.new()
    |> :queue.snoc({src, k, 0})
    |> bfs(graph)
    |> Map.get(dst, -1)
  end

  def bfs(q, graph, mp \\ %{}) do
    if :queue.is_empty(q) do
      mp
    else
      {curr, ct, price} = :queue.head(q)
      q = :queue.tail(q)

      cond do
        Map.get(mp, curr, price + 1) < price ->
          bfs(q, graph, mp)

        ct < 0 ->
          bfs(q, graph, Map.put(mp, curr, price))

        true ->
          Map.get(graph, curr, [])
          |> Enum.reduce(q, fn {dst, p}, q ->
            :queue.in({dst, ct - 1, p + price}, q)
          end)
          |> bfs(graph, Map.put(mp, curr, price))
      end
    end
  end
end

File: algorithms/medium/cheapest_flights_within_k_stops.cpp
class Solution {
public:
  int findCheapestPrice(int n, vector<vector<int>> &flights, int src, int dst,
                        int k) {
    unordered_map<int, vector<pair<int, int>>> graph;
    for (int i = 0; i < flights.size(); i++) {
      graph[flights[i][0]].push_back({flights[i][1], flights[i][2]});
    }

    queue<array<int, 3>> q;
    q.push({src, k, 0});
    vector<int> ans(n, -1);
    while (!q.empty()) {
      auto [curr, ct, price] = q.front();
      q.pop();
      if (ans[curr] > 0 && ans[curr] < price) {
        continue;
      }
      ans[curr] = price;
      if (ct < 0) {
        continue;
      }
      for (const auto &[d, p] : graph[curr]) {
        q.push({d, ct - 1, p + price});
      }
    }
    return ans[dst];
  }
};
